### PR TITLE
tests: Fetch the pod logs in failed cases

### PR DESCRIPTION
**Pull request by claudiubelu**
#### What type of PR is this?

/kind failing-test
/sig testing

#### What this PR does / why we need it:

For some test failures, checking the pod logs could potentially yield some interesting information, which could be used to further
investigate certain failures / flakes (for example, if there are some networking issues, we could at least see if requests reach the containers, (agnhost logs the connections / requests), or if there were any other issues during the container's startup).

Sample test failure: https://paste.ubuntu.com/p/FT9KBmNrjV/

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by oomichi** [comment]
Thanks for doing this.
This seems useful to investigate root causes.

/lgtm

File: test/e2e/framework/pod/resource.go [hunk 1]
```diff
@@ -388,6 +388,29 @@ func logPodTerminationMessages(pods []v1.Pod) {
...
+// logPodLogs logs the container logs from pods in the given namespace. This can be helpful for debugging
+// issues that do not cause the container to fail (e.g.: network connectivity issues)
+// We will log at most the first 5 pods. There are tests that spawn more, we don't want to log so many
+// in those cases.
+func logPodLogs(c clientset.Interface, namespace string, pods []v1.Pod) {
+	maxPods := len(pods)
+	if maxPods > 5 {
+		maxPods = 5
+	}
+
```
[the remark this answers is not in the document]
Comment (neolit123) [reply test/e2e/framework/pod/resource.go]: I agree that default to the first 5 pods in and is a bit random.

I think by default it should be 0 pods. Specific tests can opt in. I'd also add a utility function that applies the label to a Pod.
The label value can also hold how many lines to dump from a pod. -1 == all, cannot parse the int == 0

**Comment by neolit123** [comment]
Added one comment here https://github.com/kubernetes/kubernetes/pull/105451#discussion_r741009612

/unassign
We already have 2 approvers here.

File: test/e2e/framework/pod/resource.go [hunk 2]
```diff
@@ -544,28 +585,30 @@ func checkPodsCondition(c clientset.Interface, ns string, podNames []string, tim
...
 // utility function for gomega Eventually
-func getPodLogsInternal(c clientset.Interface, namespace, podName, containerName string, previous bool, sinceTime *metav1.Time) (string, error) {
+// if tailLen is -1, all the pod logs will be returned.
+func getPodLogsInternal(c clientset.Interface, namespace, podName, containerName string, previous bool, sinceTime *metav1.Time, tailLen int) (string, error) {
 	request := c.CoreV1().RESTClient().Get().
 		Resource("pods").
 		Namespace(namespace).
 		Name(podName).SubResource("log").
 		Param("container", containerName).
+		Param("tail", strconv.Itoa(tailLen)).
```
Comment (spiffxp) [inline test/e2e/framework/pod/resource.go:611]: ref: https://github.com/kubernetes/kubernetes/blob/ab609a418ea804c896cbcba4d93a77e46ca76dff/staging/src/k8s.io/api/core/v1/types.go#L5233-L5236

My read is you should be using `tailLines` and omitting it if you're passing in -1.

File: test/e2e/framework/pod/resource.go [hunk 3]
```diff
@@ -388,6 +393,41 @@ func logPodTerminationMessages(pods []v1.Pod) {
...
+
+	for i := 0; i < maxPods; i++ {
+		pod := logPods[i]
+		for _, container := range pod.Spec.Containers {
+			logs, err := getPodLogsInternal(c, namespace, pod.Name, container.Name, false, nil, 42)
+			if err != nil {
+				e2elog.Logf("unable to fetch %s/%s/%s logs: %v", pod.Namespace, pod.Name, container.Name, err)
+				continue
+			}
+			e2elog.Logf("%s/%s/%s logs:\n%s", pod.Namespace, pod.Name, container.Name, logs)
```
Comment (spiffxp) [inline test/e2e/framework/pod/resource.go:426]: I'm not sure how I feel about polluting build-log.txt with this.  What about dumping into something like `artifacts/pods/{namespace}/{pod}/{container}/logs.txt`?

Comment (claudiubelu) [reply test/e2e/framework/pod/resource.go]: Done

**Comment by claudiubelu** [comment]
Test run with this PR (test intetionally fails, so we can see the Pod logs in the failure case): https://paste.ubuntu.com/p/mRKgw3dMvs/

Sample Pod logs fetched:

```
ubuntu@ubuntu:~/run_e2e$ cat test_artifacts/kubectl-3144/agnhost-replica-69556bf845-6rvnq/replica/logs.txt
2021/11/04 16:04:47 Registering to primary: agnhost-primary:6379/register?host=10.240.0.43
2021/11/04 16:04:52 encountered error while registering to primary: Get "http://agnhost-primary:6379/register?host=10.240.0.43": context deadline exceeded (Client.Timeout exceeded while awaiting headers). Retrying in 1s.
2021/11/04 16:04:53 Registering to primary: agnhost-primary:6379/register?host=10.240.0.43
2021/11/04 16:04:58 encountered error while registering to primary: Get "http://agnhost-primary:6379/register?host=10.240.0.43": context deadline exceeded (Client.Timeout exceeded while awaiting headers). Retrying in 1s.
2021/11/04 16:04:59 Registering to primary: agnhost-primary:6379/register?host=10.240.0.43
2021/11/04 16:05:04 encountered error while registering to primary: Get "http://agnhost-primary:6379/register?host=10.240.0.43": context deadline exceeded (Client.Timeout exceeded while awaiting headers). Retrying in 1s.
2021/11/04 16:05:05 Timed out while registering to primary.
```

**Comment by claudiubelu** [comment]
@spiffxp , does this look better to you now?

/milestone v1.23
We'd like to get this in for 1.23 and this is a test only PR

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm

**Comment by spiffxp** [comment]
Sorry for the delay in review, I've been out

**Pull request merged**